### ISSUE TITLE
Name config.js something else.

### ISSUE THREAD

**Issue by Zoltu** [jspm/jspm-cli#715]
Putting a generated file named `config.js` into the root of the application is likely to run into a conflict with other tools (name is too generic).  I recommend naming it something like `jspm-config.js`.

Also, a name like `jspm-config.js` will be more apparent to the user where it comes from.  When I was first starting with JSPM, it took me a while to figure out where `config.js` was coming from.

It is possible that it is named generically because it is shared between JSPM and SystemJS, though I believe SystemJS gets the file passed in by name by the user, so SystemJS isn't tightly coupled with that name.

**Comment by lookfirst** [comment]
Name the file whatever you want and put it wherever you want. 

https://github.com/lookfirst/systemjs-seed/blob/0b338956ea81306492530cb151c08bd4baa7d9f3/package.json#L19
https://github.com/lookfirst/systemjs-seed/blob/master/system.config.js

**Comment by Zoltu** [comment]
Ah, I didn't know about this one. Thanks!  I still recommend against defaulting to such a generic name, but you can consider this a feature request rather than a bug and close as "won't add" if that is what the maintainers deem best.

**Comment by lookfirst** [comment]
fyi, in the [jspm init phase](https://github.com/jspm/jspm-cli/wiki/Getting-Started#2-create-a-project), you're asked what the name of the file is and where to put it.

**Comment by guybedford** [comment]
It's possible to consider changing the default, but will be a many-months process at this point. I was wondering if it's worth picking a name that make it clearer to users not to mind as much if it is getting large.

That said, I don't see any real problem with `config.js`.

**Comment by johnsoftek** [comment]
+1 jspm-config.js as the default.

**Comment by ArmorDarks** [comment]
+1 for more specific name, like `jspm-config.js` or simply `jspm.js`

For example, SCSS-Link using as config simple `.scss-lint.yml` (with trailing dot). Looks quite nice.

Few other popular examples: `.editorconfig`, `gruntfile.coffee`, `.gitignore`

**Comment by AlicanC** [comment]
:+1:

A file named `config.js` looks more like something unique for the project instead of a file for a specific library.

I would highly suggest against `jspm.config.js`. I don't know how developers can think using dot as a separator is a good idea.

Kebab case `jspm-config.js` looks fine as a new default.

---

Yes, this is configurable, but I don't think that fact adds anything to this discussion. Configurability doesn't make defaults acceptable. Imagine the default being `neonazihailsatan.js` and we were being against its change because it is configurable. Current default is not that bad of course, but if it can get any better, it should.

---

This should be discussed for months. I don't think this will be open to discussion after major version zero so let's end up with something solid.

**Comment by ArmorDarks** [comment]
> I would highly suggest against jspm.config.js. I don't know how developers can think using dot as a separator is a good idea.

I'm not sure is it related to my input or no. Just to clarify — I mean trailing dot, in the beginning of the file, which used in UNIX systems to mark directories and files as hidden. Usually used to for system files and directories.

**Comment by AlicanC** [comment]
It wasn't related to your input.

I figured someone will eventually come and suggest `jspm.conf.js` since some libraries like Karma uses that format. The problem with that is `.conf` doesn't serve as an extension there. `.tar.gz` means the file should be ungzipped and then treated as a tar file. `.html.slim` means the file should be parsed with Slim and then treated as an HTML file. What does `.conf.js` mean now? Should it be parsed with JS and then treated as a conf file? Where is the specification for conf files? There is none so it is misleading.

---

About your input, doesn't "trailing dot" mean a dot at the end? I think you mean "leading dot". Please correct me if I'm wrong.

I don't think leading dot is a good idea. It doesn't feel Node-y since I haven't seen any library using it. I would support it if it was making things any better, but I don't think it's.

**Comment by ArmorDarks** [comment]
> What does .conf.js mean now? Should it be parsed with JS and then treated as a conf file? Where is the specification for conf files? There is none so it is misleading.

That's sounds reasonable, so I must agree.

> About your input, doesn't "trailing dot" mean a dot at the end? I think you mean "leading dot". Please correct me if I'm wrong.

My bad :) I meant leading dot. Sorry for misleading.

> I don't think leading dot is a good idea. It doesn't feel Node-y since I haven't seen any library using it. I would support it if it was making things any better, but I don't think it is.

A some Grunt-related tasks using it, for example. I'd say, it's usually depends on author's OS — if it's UNIX-based system they just used to leading dots in config files, so they use them in their modules too.

**Comment by KenRmk** [comment]
I'd also like to see `config.js` renamed to something else. Having `configFile` option in `package.json` is nice but it only solves a problem we shouldn't even be introducing in the first place. I know it's just a single option and there's the `jspm init`, etc... but those are just excuses

**Comment by laurentgoudet** [comment]
+1, convention over configuration. jspm-config.js un nice as it starts with "js" so everyone knows immediately what it's for.

**Comment by juergenzimmermann** [comment]
+1 for jspm-config.js

**Comment by neothemachine** [comment]
I actually wondered why there is only one config.js. All the dev dependencies (used in tests etc.) get included there as well which doesn't seem right. Certainly strange to serve that in production.

**Comment by guybedford** [comment]
Yes ideally there should be a `config.dev.js` and a `config.js` as well.

**Comment by chrisdostert** [comment]
+1 for jspm-config.js

**Comment by kylecordes** [comment]
I am skeptical of the proposed name jspm-config.js; from my (admittedly limited) understanding, it is more correct to think of this file as belonging to SystemJS, and that JSPM should be thought of as mostly a tool to automate managing the contents of that file. There are even examples around the web (including the one right on the Angular.io website explaining how to use Angular2 with typescript) which use SystemJS, and use the config.js file, but do not use JSPM at all.

(However, I am eager to see the following. Documentation and examples of how this config file might live somewhere other than the root of the project. It is somewhat hard to take the tool seriously with the assumption that the root of the project is a suitable place from which to serve a file via a web server. Even in development-mode, this feels very retro.)

**Comment by guybedford** [comment]
@kylecordes `system.config.js` or `system-config.js` could certainly be a good option as well. To set the base folder, `jspm init -p` asks to set the server baseURL which corresponds to the config.js default folder.

**Comment by kylecordes** [comment]
Thank you. With that hint and finding a page or two were someone worked with a different base URL, I got it working as I had hoped. I will probably publish an example later, showing this.

I think the difficulty was that it is not especially clear in the documentation what goes where. Even as a look at the JSPM documentation now, I am unable to find anything documenting the set of options available inside the jspm key in package.json.  But the answer is easily found with the command "jspm help init", jspm init accepts the base URL right there on the command line. It takes literally one plane parameter when performing jspm set up to put the web root somewhere other than the base of a project, it could hardly be any easier.

**Comment by ArmorDarks** [comment]
Why not simple `jspm.js` or `.jspm.js`?

Or even just `.jspm` (well-known examples: `.editorconfig`, `.eslintrc`, `.gitmodules`, `.gitignore`)

**Comment by Zoltu** [comment]
It is part of a distribution, not tools configuration, so I don't believe it makes sense to be prefixed with `.` like the examples you listed.  Also, the file is generated by JSPM but it is used by SystemJS so it isn't entirely clear which one it should be named after.

**Comment by ArmorDarks** [comment]
> It is part of a distribution, not tools configuration

in such case `config` word definitely should be dropped from name. `jspm-config.js`-like examples doesn't represent reality.

And isn't `config.js` used for mapping endpoints? Otherwise it could be just temporary file.

> Also, the file is generated by JSPM but it is used by SystemJS so it isn't entirely clear which one it should be named after.

It's used by JSPM at first place and passed to SystemJS, so naming it after SystemJS quite confusing.

**Comment by guybedford** [comment]
Related to this - the package.json `configFile` property doesn't quite sound right to me as well.

I'm wondering about changing this to a generic `configFiles` property:

package.json
```json
{
  "configFiles": {
    "jspm": "jspm.custom-name.js",
    "jspm:build": "jspm.custom-build-file.json"
  }
}
```

This way it can expand to other project uses or contain _dev_ and _build_ config files for jspm as well in future.

Does this sound like it would have uses beyond jspm? Does anyone think this is a bad idea? Feedback would be appreciated.

**Comment by AlicanC** [comment]
@guybedford

> Does this sound like it would have uses beyond jspm?

What use case does it have with JSPM? I have `defaultJSExtensions`, `transpiler`, `babelOptions`, `paths` and `map` in my "config.js". Why would I want to change any of them when I'm building? (Or am I *underusing* JSPM?)

**Comment by kylecordes** [comment]
Regarding the idea of adding another top level key to the package.json file, I'm somewhat leery of that. If various projects and up doing the same thing, eventually those files will end up a tangled nest of different programs competing configuration information. I think the lightest touch approach taken by JSPM today, of adding just one single key, makes the most sense.

Also as I wrote earlier, a fundamental conundrum here is that this configuration file in one sense belongs to JSPM (which extensively manipulates it) but in another sense belongs just to SystemJS. I can imagine a future (though I'm not sure if @guybedford  imagines the same future) if some other module loader gains popularity, where JSPM could alternatively manipulate a configuration file suited for that module loader. Or correspondingly, perhaps SystemJS will become quite popular, but some other package manager will emerge that manipulates its configuration file in lieu of JSPM. In any event, I think it is a brilliant choice to implement JSPM and SystemJS separately, but it does lead to this mystery of what to call the file sitting between them.

**Comment by ArmorDarks** [comment]
`.systemjs.jspm.js` or `.jspm.systemjs.js`? So many js...

**Comment by AlicanC** [comment]
You could work on a JSPM-using project for months and not realize it uses SystemJS if you are not the guy who actually wrote the `<script>` tags.

This file is a lot more of a JSPM thing than a SystemJS thing. Putting SystemJS in this file's name would be like naming all of your JavaScript files "blah-blah-browser.js" because they are for browsers.

**Comment by glen-84** [comment]
This is tricky, but I think that at the end of the day, this is a file for SystemJS configuration, which may be manipulated by tools other than JSPM.

... or should these tools use a separate file for their own SystemJS configuration?

**Comment by AlicanC** [comment]
@glen-84, you can't even build JSPM projects with SystemJS. Any other tool that expects 100% SystemJS compatibility won't be satisfied by this file.

**Comment by glen-84** [comment]
@AlicanC I don't understand what you mean. :worried:

**Comment by AlicanC** [comment]
@glen-84 you (plural) see `config.js` as something that JSPM creates for SystemJS, but SystemJS can't create a bundle using `config.js` created by JSPM. (You need `jspm.Builder` for that.) So this file can't be for SystemJS, because it doesn't fully work with SystemJS.

And any JSPM alternative would have three options:
- Generate 100% SystemJS compatible file.
- Generate another incompatible file.
- Generate JSPM-compatible file that works with JSPM builder for bundling and SystemJS on runtime.

If JSPM and SystemJS can decide on a format that doesn't require a custom builder like JSPM's builder, than you could call `config.js` a file for SystemJS.

**Comment by glen-84** [comment]
@AlicanC Are you saying that the `System` in `config.js` uses a JSPM-specific loader?

I always thought that JSPM was developed as an extension to SystemJS.

**Comment by AlicanC** [comment]
@glen-84 It does not use a JSPM-specific loader. The loader which you use for development is 100% SystemJS. The builder which you use to create bundles for production is JSPM-specific. AFAICR, the JSPM builder is just a small wrapper that does small modifications to the configuration so the actual SystemJS builder can use it.

So we can say that `config.js` is 100% compatible with SystemJS runtime and like 95% compatible with SystemJS builder.

**Comment by glen-84** [comment]
Doesn't make any sense to me.

Anyway, why would Guy [suggest](https://github.com/jspm/jspm-cli/issues/715#issuecomment-144020100) the possibility of using `system-config.js` if it wasn't appropriate?

**Comment by AlicanC** [comment]
I would like to not speculate on why Guy does the things he does. All we do here is just bikeshedding anyways. Anything other than `config.js` is fine.

**Comment by guybedford** [comment]
Released in jspm 0.17 beta.
